### ISSUE TITLE
Confusing error with bad generics in structs "error: expected one of `!`, `(`, `,`, `=`, `>`, `?`, `for`, `~`, lifetime, or path, found keyword `fn`"

### ISSUE THREAD

**Issue by asquared31415** [rust-lang/rust#92529]
Given the following code: 

```rust
// Only the *first* error of this kind gets reported
struct FnS<F: fn()> {
    f: F
}

struct InS<I: in> {
    i: I
}

// works like `union` in edition 2015, errors as keywords do in 2018+
struct AsyncS<A: async> {
    a: A
}

// WORKS
struct UnionS<U: union> {
    u: U
}
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5065c4a01d77d7e17aac442b1f75b7b2)

For structs with a keyword in place of a trait in generics an error of the form

```
error: expected one of `!`, `(`, `,`, `=`, `>`, `?`, `for`, `~`, lifetime, or path, found keyword `fn`
 --> src/lib.rs:2:15
  |
2 | struct FnS<F: fn()> {
  |               ^^ expected one of 10 possible tokens
```
appears. This happens for all keywords, but does not happen for "weak" keywords like `union` or keywords like `async` if compiled in an edition where they were not yet keywords.

Additionally **only the first** of such errors encountered is reported.

I'm not sure what a more clear diagnostic would be for the general case.  However for the `fn` keyword case specifically it should probably include a `help` to use an `Fn` *trait* instead.

This occurs across all compiler versions and editions (barring keyword changes)
Stable 1.57.0
Beta 1.58.0-beta.2
Nightly 1.59.0-nightly 2022-01-02 (8f3238f898163f09726c)

**Comment by estebank** [comment]
Current output:

```
error: expected identifier, found keyword `fn`
 --> src/lib.rs:2:15
  |
2 | struct FnS<F: fn()> {
  |               ^^
  |
help: use `Fn` to refer to the trait
  |
2 | struct FnS<F: Fn()> {
  |               ~~

error: expected identifier, found keyword `in`
 --> src/lib.rs:6:15
  |
6 | struct InS<I: in> {
  |               ^^ expected identifier, found keyword

error: expected identifier, found keyword `async`
  --> src/lib.rs:11:18
   |
11 | struct AsyncS<A: async> {
   |                  ^^^^^ expected identifier, found keyword

error[E0405]: cannot find trait `r#in` in this scope
 --> src/lib.rs:6:15
  |
6 | struct InS<I: in> {
  |               ^^ help: a trait with a similar name exists: `Fn`
 --> /rustc/270c94e484e19764a2832ef918c95224eb3f17c7/library/core/src/ops/function.rs:158:1
  |
  = note: similarly named trait `Fn` defined here

error[E0405]: cannot find trait `r#async` in this scope
  --> src/lib.rs:11:18
   |
11 | struct AsyncS<A: async> {
   |                  ^^^^^ not found in this scope

error[E0405]: cannot find trait `union` in this scope
  --> src/lib.rs:16:18
   |
16 | struct UnionS<U: union> {
   |                  ^^^^^ not found in this scope
```

I believe that we have addressed the original ask. Feel free to reopen if you believe there's more work needed.
